File: 09/solution.ts
const input = await Bun.file(`${import.meta.dir}/input.txt`).text();
const lines = input.trim().split("\n");

// brute force
function part1() {
	const points = lines.map((l) => l.split(",").map(Number));
	const areas: number[] = [];

	for (let i = 0; i < points.length; i++) {
		for (let k = i; k < points.length; k++) {
			const [ax, ay] = points[i];
			const [bx, by] = points[k];
			areas.push((Math.abs(ax - bx) + 1) * (Math.abs(ay - by) + 1));
		}
	}

	return Math.max(...areas);
}

function isInsideOrOnBoundary(
	polygon: number[][],
	px: number,
	py: number,
): boolean {
	let crossings = 0;

	for (let i = 0; i < polygon.length; i++) {
		const [x1, y1] = polygon[i];
		const [x2, y2] = polygon[(i + 1) % polygon.length];

		// on vertical edge?
		if (
			x1 === x2 &&
			px === x1 &&
			py >= Math.min(y1, y2) &&
			py <= Math.max(y1, y2)
		)
			return true;
		// on horizontal edge?
		if (
			y1 === y2 &&
			py === y1 &&
			px >= Math.min(x1, x2) &&
			px <= Math.max(x1, x2)
		)
			return true;
		// ray casting (vertical edges only)
		if (x1 === x2 && x1 > px && py > Math.min(y1, y2) && py <= Math.max(y1, y2))
			crossings++;
	}

	return crossings % 2 === 1;
}

function rectangleFullyInside(
	polygon: number[][],
	ax: number,
	ay: number,
	bx: number,
	by: number,
): boolean {
	const minX = Math.min(ax, bx);
	const maxX = Math.max(ax, bx);
	const minY = Math.min(ay, by);
	const maxY = Math.max(ay, by);

	// check other 2 corners are inside
	if (!isInsideOrOnBoundary(polygon, minX, maxY)) return false;
	if (!isInsideOrOnBoundary(polygon, maxX, minY)) return false;

	// check no edge cuts through
	for (let i = 0; i < polygon.length; i++) {
		const [x1, y1] = polygon[i];
		const [x2, y2] = polygon[(i + 1) % polygon.length];

		// horizontal edge cuts through?
		if (
			y1 === y2 &&
			y1 > minY &&
			y1 < maxY &&
			Math.min(x1, x2) < maxX &&
			Math.max(x1, x2) > minX
		)
			return false;
		// vertical edge cuts through?
		if (
			x1 === x2 &&
			x1 > minX &&
			x1 < maxX &&
			Math.min(y1, y2) < maxY &&
			Math.max(y1, y2) > minY
		)
			return false;
	}

	return true;
}

function part2() {
	const points = lines.map((l) => l.split(",").map(Number));
	const areas: number[] = [];

	for (let i = 0; i < points.length; i++) {
		for (let k = i + 1; k < points.length; k++) {
			const [ax, ay] = points[i];
			const [bx, by] = points[k];

			if (rectangleFullyInside(points, ax, ay, bx, by)) {
				areas.push((Math.abs(ax - bx) + 1) * (Math.abs(ay - by) + 1));
			}
		}
	}

	return Math.max(...areas);
}

console.log("Part 1:", part1());
console.log("Part 2:", part2());

File: 09/solution.ml
open Base
open Stdio

let read_input () = In_channel.read_lines "09/input.txt"

let parse_point line =
  let x, y = String.lsplit2_exn line ~on:',' in
  (Int.of_string x, Int.of_string y)

let part1 points =
  List.cartesian_product points points
  |> List.map ~f:(fun ((ax, ay), (bx, by)) ->
      (Int.abs (ax - bx) + 1) * (Int.abs (ay - by) + 1))
  |> List.fold ~init:0 ~f:Int.max

let is_inside_or_on_boundary polygon_arr n (px, py) =
  let crossings = ref 0 in
  let on_edge = ref false in
  for i = 0 to n - 1 do
    let ax, ay = polygon_arr.(i) in
    let bx, by = polygon_arr.((i + 1) % n) in
    let min_x, max_x = (min ax bx, max ax bx) in
    let min_y, max_y = (min ay by, max ay by) in
    if ax = bx && px = ax && py >= min_y && py <= max_y then on_edge := true;
    if ay = by && py = ay && px >= min_x && px <= max_x then on_edge := true;
    if ax = bx && ax > px && py > min_y && py <= max_y then Int.incr crossings
  done;
  !on_edge || !crossings % 2 = 1

let rectangle_fully_inside polygon_arr n (ax, ay) (bx, by) =
  let min_x, max_x = (min ax bx, max ax bx) in
  let min_y, max_y = (min ay by, max ay by) in
  if not (is_inside_or_on_boundary polygon_arr n (min_x, max_y)) then false
  else if not (is_inside_or_on_boundary polygon_arr n (max_x, min_y)) then false
  else
    let cuts_through = ref false in
    for i = 0 to n - 1 do
      let x1, y1 = polygon_arr.(i) in
      let x2, y2 = polygon_arr.((i + 1) % n) in
      if
        y1 = y2 && y1 > min_y && y1 < max_y
        && min x1 x2 < max_x
        && max x1 x2 > min_x
      then cuts_through := true;
      if
        x1 = x2 && x1 > min_x && x1 < max_x
        && min y1 y2 < max_y
        && max y1 y2 > min_y
      then cuts_through := true
    done;
    not !cuts_through

let part2 points =
  let polygon_arr = Array.of_list points in
  let n = Array.length polygon_arr in
  List.cartesian_product points points
  |> List.filter ~f:(fun (p1, p2) -> rectangle_fully_inside polygon_arr n p1 p2)
  |> List.map ~f:(fun ((ax, ay), (bx, by)) ->
      (Int.abs (ax - bx) + 1) * (Int.abs (ay - by) + 1))
  |> List.fold ~init:0 ~f:Int.max

let () =
  let points = read_input () |> List.map ~f:parse_point in
  printf "Part 1: %d\n" (part1 points);
  printf "Part 2: %d\n" (part2 points)
